### ISSUE TITLE
Contrasting text outline, shadow or border

### ISSUE THREAD

**Issue by jstawik** [flameshot-org/flameshot#1249]
It's very important for me for the text to be clearly visible when added to a screenshot. This is usually easily achieved by using black text with white background or white text with black background - but other colours usually look good with either (except maybe for very dark and very light colours).
I know that usually the implementations involve:
- Creating a shadow
- Creating a border

I do realise I can put a rectangle under text, but it's ugly and obscures a lot more of the image.

As a humorous reference: :smile: ![white text with black outline can be read on any color](https://2.bp.blogspot.com/-TGEt81O_qaI/Uxns70ZKVEI/AAAAAAAACE4/V5v81OF9sT4/s1600/white-text-black-outline.png)

**Comment by borgmanJeremy** [comment]
Seem's straightforward enough. I'd like this to be integrated into the side menu with the other font effects (bold, strikethrough, etc).

I want to focus on other issues right now but I'll happily merge this. Here is some documentation from QT on how to do it: https://doc.qt.io/qt-5/qgraphicsdropshadoweffect.html

**Comment by Samuel-Martin23** [comment]
So I decided to take a swing at this. While I mostly code in C, I have worked with C++ and Qt before. Implementing the effect shouldn't be too hard, but my main confusion comes from the moc_textconfig.cpp file. It seems all of the text code effects are written there, so I should probably put my border code in there as well. Is there documentation on this file or something I could look at? For me, it's just a little bit confusing what's going on.

**Comment by Samuel-Martin23** [comment]
@borgmanJeremy I just whipped up this little demo, and idk if the shadow effect would work super well. Maybe some other effect?
![demo](https://user-images.githubusercontent.com/59482608/112888389-245e2100-9089-11eb-9d82-0a2449ac5431.gif)

**Comment by jstawik** [comment]
@Samuel-Martin23 thanks for taking up this task, as the original requestor this means a lot to me.

If I can throw a suggestion from my side - it's important to keep in mind that the shadow is just a mean to obtain what I've showcased in the original post - namely the visibility independent of the background. That said I'd consider making the shadow a bit less blurry (more opaque), and more "centered" under the text.
Thanks again for working on this issue :)

**Comment by Samuel-Martin23** [comment]
Ok, I think the biggest thing for me is if you make the blur too big the letters look like they are glowing instead of having an outline/border. 

<img width="576" alt="Screen Shot 2021-03-29 at 12 52 59 PM" src="https://user-images.githubusercontent.com/59482608/112892425-24145480-908e-11eb-8a36-3c98682a9a10.png">

**Comment by jstawik** [comment]
You would have to make `blurRadius` 1 or 0 but at the same time somehow achieve the shadow being bigger than the object casting it I think

**Comment by Samuel-Martin23** [comment]
So create two shadows then? One being the color and the other being a black shadow?

**Comment by jstawik** [comment]
I don't think we need the colour. Only a black. opaque shadow with minimal blur, centered.

**Comment by Samuel-Martin23** [comment]
The top one is just normal text and the bottom one has the black blur on it.
<img width="120" alt="Screen Shot 2021-03-29 at 4 32 38 PM" src="https://user-images.githubusercontent.com/59482608/112912500-7f553f80-90ac-11eb-8f96-f0c4ff45cb0b.png">

**Comment by snide** [comment]
Hey folks. I'm a designer by trade and spend a lot of time with color theory and calculations. To make sure the colors always works you'll want to apply both a stroke and a shadow. You should decide which color to apply (black or white) as the stroke based upon the luminance of the color of the text. Here's a way to [calculate luminance](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef). The stroke color can differ, but the shadow color is OK to always be a level of black.

![image](https://user-images.githubusercontent.com/324519/112924230-9b64db00-90c4-11eb-98e4-1878ec17eee5.png)

**Comment by Samuel-Martin23** [comment]
Gotcha, idk if the current effects lib has a stoke effect. I'll look around with the style sheets maybe?

**Comment by snide** [comment]
If you have the ability to apply multiple shadows, you can likely fake a stroke by applying two shadows. One with a 0 blur radius to mimic the stroke, and then another with a higher value for the soft shadow.

I don't know much about QT (I'm a CSS nerd) but if you need help fiddling once you get it in PR I'm happy to help.

TY SO MUCH for attacking this one. This iwill make me so excited

**Comment by Samuel-Martin23** [comment]
I don't think you can combine graphics effects tho. Also, thank you all for helping me walk through this.

**Comment by borgmanJeremy** [comment]
@snide @Samuel-Martin23 Thanks for working on this!

By the way we have a luminosity calculation here: https://github.com/flameshot-org/flameshot/blob/71d595fbcea3eac9d1aaa51f42cb53935bd9f3a7/src/utils/colorutils.cpp#L11

We currently use it to figure out if the text in the incrementing bubbles should be white or black.

If you need any help or guidance from me let me know!

**Comment by Samuel-Martin23** [comment]
Thanks, @borgmanJeremy! I have been looking around to see if I could use the Qpainter effect for the text, but I don't think it will work. We have the black shadow effect recommend by @snide but we need the border/stroke effect around the text.

**Comment by mikkovedru** [comment]
To me, this is the only feature that I miss. The readability of the text is quite terrible, especially when using the red color - the one I use almost exclusively.
![image](https://user-images.githubusercontent.com/2521942/157039065-ffd95402-454c-4301-ac80-f512c098e9fa.png)

The same contrasting effect should also be optionally used at least with arrows.

---
Another helpful thing to know is the color wheel and the way complementary (high contrast and high impact; together, these colors will appear brighter and more prominent) colors.

Here is a short and nice overview of the color wheel:
https://www.canva.com/colors/color-wheel/

**Comment by Mobuos** [comment]
This is also something I miss a lot from other tools (Like ShareX), in general the drawing elements (Arrows, text, rectangles...) don't look too good if the background isn't a solid color...

**Comment by derek-shnosh** [comment]
Heya, just started using Flameshot and stumbled upon this issue after noticing there's no stroke/border option for text. Curious if this is still under consideration.

**Comment by matti** [comment]
wtf guys

**Comment by matti** [comment]
![Image](https://github.com/user-attachments/assets/3a24d75d-c878-4201-8dc7-efc0022e4fa9)

skitch from 2014 has this, but it barely works in 2024 anymore

**Comment by matti** [comment]
![Image](https://github.com/user-attachments/assets/16651a56-b534-412e-860c-57aa727a7a0e)

this app promises that but it's not there in the real app :DDDd

**Comment by matti** [comment]
<img width="863" alt="Image" src="https://github.com/user-attachments/assets/2dcda016-5b5a-4dfa-878e-7206d46c6ab0" />

best so far
